direct/tpm2/hash_sequence_hash_test.go

```diff
@@ -0,0 +1,233 @@
+package tpm2
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"math/rand"
+	"testing"
+
+	"github.com/google/go-tpm/direct/structures/tpm"
+	"github.com/google/go-tpm/direct/structures/tpm2b"
+	"github.com/google/go-tpm/direct/transport"
+	"github.com/google/go-tpm/direct/transport/simulator"
+)
+
+func TestHash(t *testing.T) {
+
+	thetpm, err := simulator.OpenSimulator()
+	if err != nil {
+		t.Fatalf("could not connect to TPM simulator: %v", err)
+	}
+	defer thetpm.Close()
+
+	run := func(t *testing.T, data []byte, hierarchy tpm.Handle, thetpm transport.TPM) {
+		hash := Hash{
+			Data:      tpm2b.MaxBuffer{Buffer: data},
+			HashAlg:   tpm.AlgSHA256,
+			Hierarchy: hierarchy,
+		}
+		rspHash, err := hash.Execute(thetpm)
+		if err != nil {
+			t.Fatalf("Hash failed: %v", err)
+		}
+		gotDigest := rspHash.OutHash.Buffer
+		wantDigest := sha256.Sum256(data)
+
+		if !bytes.Equal(gotDigest, wantDigest[:]) {
+			t.Errorf("Hash(%q) returned digest %x, want %x", data, gotDigest, wantDigest)
+		}
+	}
+
+	t.Run("Null hierarchy", func(t *testing.T) {
+		run(t, []byte("fiona"), tpm.RHNull, thetpm)
+	})
+
+	t.Run("Owner hierarchy", func(t *testing.T) {
+		run(t, []byte("charlie"), tpm.RHOwner, thetpm)
+	})
+}
+
+func TestHashNullHierarchy(t *testing.T) {
+	thetpm, err := simulator.OpenSimulator()
+	if err != nil {
+		t.Fatalf("could not connect to TPM simulator: %v", err)
+	}
+	defer thetpm.Close()
+
+	data := []byte("carolyn")
+	hash := Hash{
+		Data:    tpm2b.MaxBuffer{Buffer: data},
+		HashAlg: tpm.AlgSHA256,
+	}
+	rspHash, err := hash.Execute(thetpm)
+	if err != nil {
+		t.Fatalf("Hash failed: %v", err)
+	}
+	gotDigest := rspHash.OutHash.Buffer
+	wantDigest := sha256.Sum256(data)
+
+	if !bytes.Equal(gotDigest, wantDigest[:]) {
+		t.Errorf("Hash(%q) returned digest %x, want %x", data, gotDigest, wantDigest)
+	}
+}
+
+func TestHashSequence(t *testing.T) {
+	thetpm, err := simulator.OpenSimulator()
+	if err != nil {
+		t.Fatalf("could not connect to TPM simulator: %v", err)
+	}
+	defer thetpm.Close()
+
+	run := func(t *testing.T, bufferSize int, password string, hierarchy tpm.Handle, thetpm transport.TPM) {
+		maxDigestBuffer := 1024
```

Comment:
a TPM2B_MAX_BUFFER has maximum size "MAX_DIGEST" which can be read from the TPM as a "PT_MAX_DIGEST" and is always at least 1024 (so this is safe)

Reply:
Acknowledged!